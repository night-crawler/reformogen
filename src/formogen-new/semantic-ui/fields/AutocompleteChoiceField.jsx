import PropTypes from 'prop-types';
import React from 'react';

import _ from 'lodash';

import { Form } from 'semantic-ui-react';
import Select from 'react-select';

import { errorsType, layoutOptsType } from '../../fieldPropTypes';
import Label from '../common/Label';
import ErrorsList from '../common/ErrorsList';


const makeReactSelectOptions = (choices) => {
    return choices.map(([key, val]) => {
        return {
            value: key,
            label: val,
        };
    });
};


AutocompleteChoiceField.propTypes = {
    type: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    choices: PropTypes.arrayOf(PropTypes.array),
    max_length: PropTypes.number,
    help_text: PropTypes.string.isRequired,
    placeholder: PropTypes.string,
    errors: errorsType,

    required: PropTypes.bool,
    editable: PropTypes.bool,

    helpTextOnHover: PropTypes.bool,
    layoutOpts: layoutOptsType,

    updateProps: PropTypes.func,
    onChange: PropTypes.func,
};
export default function AutocompleteChoiceField(props) {
    const handleChange = (newVal) => {
        props.onChange(
            null,
            { name: props.name, value: newVal ? newVal.value : null }
        );
    };

    let _props = {
        clearable: !props.required,
        name: props.name,
        value: props.value,
        placeholder: props.placeholder,
        options: makeReactSelectOptions(props.choices),
        onChange: handleChange,
        inputProps: { type: 'react-type' },  // fixing breaking semantic markup
    };

    if (_.isFunction(props.updateProps)) {
        _props = props.updateProps(_props, props);
    }

    return (
        <Form.Field
            required={ props.required }
            disabled={ !props.editable }
            width={ props.layoutOpts.width }
            error={ !_.isEmpty(props.errors) }
        >
            <Label { ...props } />
            <Select { ..._props } />
            {
                !props.helpTextOnHover
                    ? <span className='help-text'>{ props.help_text }</span>
                    : ''
            }
            <ErrorsList messages={ props.errors } />
        </Form.Field>
    );
}
