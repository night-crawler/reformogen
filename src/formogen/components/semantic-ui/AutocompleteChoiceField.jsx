import React from 'react';
import { Form } from 'semantic-ui-react';
import Select from 'react-select';
import _ from 'lodash';
import Label from './Label';

import propTypes from '../fieldPropTypes';


const makeReactSelectOptions = (choices) => {
    return choices.map( ([ key, val ]) => {
        return {
            value: key,
            label: val,
        };
    });
};


AutocompleteChoiceField.propTypes = propTypes;
export default function AutocompleteChoiceField(props) {
    /*
        "name": "state",
        "verbose_name": "state",
        "help_text": "",
        "blank": false,
        "null": false,
        "editable": true,
        "type": "PositiveSmallIntegerField",
        "required": true,
        "default": 255,
        "choices": [
            [0, "dead"],
            [255, "alive"],
            [30, "dried"]
        ]
     */

    const handleChange = (newVal) => {
        props.onChange(
            null,
            {name: props.name, value: newVal ? newVal.value : null }
        );
    };

    let _props = {
        clearable: !props.required,
        name: props.name,
        value: props.value,
        options: makeReactSelectOptions(props.choices),
        onChange: handleChange,
        inputProps: {type: 'react-type'},  // fixing breaking semantic markup
    };

    if (_.isFunction(props.updateProps)) {
        _props = props.updateProps(_props, props);
    }

    return (
        <Form.Field required={ props.required } disabled={ !props.editable } width={ props.width }>
            <Label { ...props } />
            <Select { ..._props } />

            { !props.helpTextOnHover ? <span className="help-text">{ props.help_text }</span> : ''}
        </Form.Field>
    );
}
