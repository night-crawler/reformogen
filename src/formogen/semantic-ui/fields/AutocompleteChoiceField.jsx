import PropTypes from 'prop-types';
import React from 'react';
import { isEmpty } from 'lodash';
import { Form } from 'semantic-ui-react';
import Select from 'react-select';

import { errorsType, layoutOptsType } from '../../fieldPropTypes';
import { FieldLabel } from '../FieldLabel';
import { ErrorsList } from '../ErrorsList';


AutocompleteChoiceField.propTypes = {
  type: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  choices: PropTypes.arrayOf(PropTypes.array),
  max_length: PropTypes.number,
  help_text: PropTypes.string,
  placeholder: PropTypes.string,
  errors: errorsType,

  required: PropTypes.bool,
  editable: PropTypes.bool,

  helpTextOnHover: PropTypes.bool,
  layoutOpts: layoutOptsType,

  getOptionLabel: PropTypes.func,
  getOptionValue: PropTypes.func,

  onChange: PropTypes.func,
};
AutocompleteChoiceField.defaultProps = {
  choices: [],
  getOptionLabel: ([ , name ]) => name,
  getOptionValue: ([ id ]) => id,
};
export function AutocompleteChoiceField(props) {
  const handleChange = ([ id ]) => props.onChange(
    null,
    { name: props.name, value: id !== undefined ? id : null }
  );

  return (
    <Form.Field
      required={ props.required }
      disabled={ !props.editable }
      width={ props.layoutOpts.width }
      error={ !isEmpty(props.errors) }
    >
      <FieldLabel { ...props } />
      <Select 
        isMulti={ false }
        clearable={ !props.required }
        name={ props.name }
        value={ props.choices.filter(([ id ]) => props.value === id ) }
        placeholder={ props.placeholder }
        options={ props.choices }
        onChange={ handleChange }

        getOptionLabel={ props.getOptionLabel }
        getOptionValue={ props.getOptionValue }
      />
      { !props.helpTextOnHover 
        ? <span className='help-text'>{ props.help_text }</span>
        : ''
      }
      <ErrorsList messages={ props.errors } />
    </Form.Field>
  );
}
