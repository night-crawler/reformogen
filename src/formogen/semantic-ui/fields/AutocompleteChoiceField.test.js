import { AutocompleteChoiceField } from './AutocompleteChoiceField';


it('<AutocompleteChoiceField />', () => {
  const wrapper = shallow(
    <AutocompleteChoiceField
      type='AutocompleteChoiceField'
      name='AutocompleteChoiceField: name'
      verbose_name='AutocompleteChoiceField: verbose_name'
      help_text='AutocompleteChoiceField: help_text'
      displayOptions={ { width: 1 } }
      choices={ [ [1, 'sample'] ] }
      value={ 1 }
    />
  );
  expect(wrapper).toMatchSnapshot();
});
