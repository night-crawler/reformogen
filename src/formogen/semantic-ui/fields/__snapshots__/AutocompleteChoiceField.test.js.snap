// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AutocompleteChoiceField /> 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AutocompleteChoiceField
    choices={
      Array [
        Array [
          1,
          "sample",
        ],
      ]
    }
    displayOptions={
      Object {
        "width": 1,
      }
    }
    getOptionLabel={[Function]}
    getOptionValue={[Function]}
    help_text="AutocompleteChoiceField: help_text"
    name="AutocompleteChoiceField: name"
    type="AutocompleteChoiceField"
    value={1}
    verbose_name="AutocompleteChoiceField: verbose_name"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <FieldLabel
          choices={
            Array [
              Array [
                1,
                "sample",
              ],
            ]
          }
          displayOptions={
            Object {
              "width": 1,
            }
          }
          getOptionLabel={[Function]}
          getOptionValue={[Function]}
          help_text="AutocompleteChoiceField: help_text"
          name="AutocompleteChoiceField: name"
          type="AutocompleteChoiceField"
          value={1}
          verbose_name="AutocompleteChoiceField: verbose_name"
        />,
        <StateManager
          clearable={true}
          defaultInputValue=""
          defaultMenuIsOpen={false}
          defaultValue={null}
          getOptionLabel={[Function]}
          getOptionValue={[Function]}
          isMulti={false}
          name="AutocompleteChoiceField: name"
          onChange={[Function]}
          options={
            Array [
              Array [
                1,
                "sample",
              ],
            ]
          }
          value={
            Array [
              Array [
                1,
                "sample",
              ],
            ]
          }
        />,
        <span
          className="help-text"
        >
          AutocompleteChoiceField: help_text
        </span>,
        <ErrorsList
          color="red"
          header="Errors"
        />,
      ],
      "disabled": true,
      "error": false,
      "required": undefined,
      "width": 1,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "choices": Array [
            Array [
              1,
              "sample",
            ],
          ],
          "displayOptions": Object {
            "width": 1,
          },
          "getOptionLabel": [Function],
          "getOptionValue": [Function],
          "help_text": "AutocompleteChoiceField: help_text",
          "name": "AutocompleteChoiceField: name",
          "type": "AutocompleteChoiceField",
          "value": 1,
          "verbose_name": "AutocompleteChoiceField: verbose_name",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "clearable": true,
          "defaultInputValue": "",
          "defaultMenuIsOpen": false,
          "defaultValue": null,
          "getOptionLabel": [Function],
          "getOptionValue": [Function],
          "isMulti": false,
          "name": "AutocompleteChoiceField: name",
          "onChange": [Function],
          "options": Array [
            Array [
              1,
              "sample",
            ],
          ],
          "placeholder": undefined,
          "value": Array [
            Array [
              1,
              "sample",
            ],
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "AutocompleteChoiceField: help_text",
          "className": "help-text",
        },
        "ref": null,
        "rendered": "AutocompleteChoiceField: help_text",
        "type": "span",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "color": "red",
          "header": "Errors",
          "messages": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <FieldLabel
            choices={
              Array [
                Array [
                  1,
                  "sample",
                ],
              ]
            }
            displayOptions={
              Object {
                "width": 1,
              }
            }
            getOptionLabel={[Function]}
            getOptionValue={[Function]}
            help_text="AutocompleteChoiceField: help_text"
            name="AutocompleteChoiceField: name"
            type="AutocompleteChoiceField"
            value={1}
            verbose_name="AutocompleteChoiceField: verbose_name"
          />,
          <StateManager
            clearable={true}
            defaultInputValue=""
            defaultMenuIsOpen={false}
            defaultValue={null}
            getOptionLabel={[Function]}
            getOptionValue={[Function]}
            isMulti={false}
            name="AutocompleteChoiceField: name"
            onChange={[Function]}
            options={
              Array [
                Array [
                  1,
                  "sample",
                ],
              ]
            }
            value={
              Array [
                Array [
                  1,
                  "sample",
                ],
              ]
            }
          />,
          <span
            className="help-text"
          >
            AutocompleteChoiceField: help_text
          </span>,
          <ErrorsList
            color="red"
            header="Errors"
          />,
        ],
        "disabled": true,
        "error": false,
        "required": undefined,
        "width": 1,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "choices": Array [
              Array [
                1,
                "sample",
              ],
            ],
            "displayOptions": Object {
              "width": 1,
            },
            "getOptionLabel": [Function],
            "getOptionValue": [Function],
            "help_text": "AutocompleteChoiceField: help_text",
            "name": "AutocompleteChoiceField: name",
            "type": "AutocompleteChoiceField",
            "value": 1,
            "verbose_name": "AutocompleteChoiceField: verbose_name",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "clearable": true,
            "defaultInputValue": "",
            "defaultMenuIsOpen": false,
            "defaultValue": null,
            "getOptionLabel": [Function],
            "getOptionValue": [Function],
            "isMulti": false,
            "name": "AutocompleteChoiceField: name",
            "onChange": [Function],
            "options": Array [
              Array [
                1,
                "sample",
              ],
            ],
            "placeholder": undefined,
            "value": Array [
              Array [
                1,
                "sample",
              ],
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "AutocompleteChoiceField: help_text",
            "className": "help-text",
          },
          "ref": null,
          "rendered": "AutocompleteChoiceField: help_text",
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "color": "red",
            "header": "Errors",
            "messages": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
